### DISCUSSION TITLE
Notebook 07 - Lesson 217 to train() function writer object has no attribute "add_scalars"

### DISCUSSION THREAD

**Discussion by AxisMeetsWorld** [mrdbourke/pytorch-deep-learning#576] [Q&A]
Hello @mrdbourke (or if there is anyone else I can ping),
           Ok, I am at the above mentioned lesson, and tried what I could to troubleshoot this one. I did some searches that didn't seem relevant, and even copied and pasted in code from the book lesson of this lecture. I noticed in this current for loop that references the `train()` function that it gets to that portion of the train function where you had set up that writer object. I am getting the above error message, and I copied in what I am seeing below as well. What is **very odd considering the way Colab is behaving** is when I go up to where we defined the `train()` function, and enter in the code `writer.` the auto population shows the "add_scalars" function as an option. Why am I getting this error if that is recognized by the code editor? It is a very confusing error message here, and I tried to look at all of the variants we defined, to make sure they were appropriately in place. Please let me know what could be the cause of this error when you get the chance. Here is what I am seeing:
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/29417743/4b6b2423-61d3-4f3d-9e7c-ff0a11324c9b)

**Comment by mrdbourke** [comment]
Hi @AxisMeetsWorld ,

This is strange. 

Do you know which notebook you're on? 

Looks like the `writer` object is being interpreted as a `'str'` object not a `SummaryWriter` object.

You could try printing `print(type(writer))` to see what the `writer` is being created as?

It should be something like: 

```python
from torch.utils.tensorboard import SummaryWriter

# Create a writer with all default settings
writer = SummaryWriter()

print(type(writer))
```

Output: 

```python
<class 'torch.utils.tensorboard.writer.SummaryWriter'>
```

The `SummaryWriter` class should definitely have the `add_scalars` method: https://pytorch.org/docs/stable/tensorboard.html#torch.utils.tensorboard.writer.SummaryWriter.add_scalars

**Reply by AxisMeetsWorld** [reply]
Unfortunately @mrdbourke, 
          That didn't turn over any clues:
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/29417743/fca24bf5-07bd-4f17-90ee-ac0729bb380a)

😞 Can't seem to figure out what was done differently than you either. Here is the code in that train section:
`# Train the target model with target dataloader to track experiments
      # note we are using train() rather than engine.train()
      train(model=model,
            train_dataloader=train_dataloader,
            test_dataloader=test_dataloader,
            optimizer=optimizer,
            loss_fn=loss_fn,
            epochs=epochs,
            device=device,
            writer=create_writer(experiment_name=dataloader_name,
                                       model_name=model_name,
                                       extra=f"{epochs}_epochs"))
`
And I also checked that area in section 4 where that code for the "train()" was created and that looked to be exactly as you have it:
`def train(model: torch.nn.Module,
          train_dataloader: torch.utils.data.DataLoader,
          test_dataloader: torch.utils.data.DataLoader,
          optimizer: torch.optim.Optimizer,
          loss_fn: torch.nn.Module,
          epochs: int,
          device: torch.device) -> Dict[str, List]:`

**I am at a loss for where to look.**
